core/src/workflows/author_entry.rs

```diff
@@ -129,31 +127,150 @@ pub mod tests {
+
+        // get the header from the top of Jill's chain
+        let state = &context1.state().unwrap();
+        let header = state.get_headers(entry_address)
+            .expect("Could not retrieve headers from authors chain")
+            .into_iter()
+            .next()
+            .expect("No headers were found for this entry in the authors chain");
+        let header_entry = Entry::ChainHeader(header);
+
+        // try and load it by its address as Jack. This means it has been communicated over the mock network
+        let mut entry: Option<Entry> = None;
+        let mut tries = 0;
+        while entry.is_none() && tries < 10 {
+            tries = tries + 1;
+            {
+                entry = get_entry_from_dht(&context2, &header_entry.address()).expect("Could not retrieve entry from DHT");
+            }
+            println!("Try {}: {:?}", tries, entry);
+            if entry.is_none() {
+                thread::sleep(time::Duration::from_millis(1000));
+            }
+        }
         assert_eq!(
-            x,
-            "{\"App\":[\"testEntryType\",\"{\\\"stuff\\\":\\\"test entry value\\\"}\"]}"
-                .to_string(),
+            entry,
+            Some(header_entry),
         );
     }
+
+
+    #[test]
+    /// test that all headers are published so an agents local chain can be reconstructed by another agent
+    fn test_reconstruct_chain_via_published_headers() {
+        let mut dna = test_dna();
+        dna.uuid = "test_reconstruct_chain_via_published_headers".to_string();
+        let netname = Some("test_reconstruct_chain_via_published_headers, the network");
+        // the ordering of these is important. Jack will get Jills DNA and AgentId headers but not visa-versa
+        let (_instance2, context2) = instance_by_name("jack", dna.clone(), netname);
+        let (_instance1, context1) = instance_by_name("jill", dna.clone(), netname);
+
+        // Jill publishes an entry
+        context1
+            .block_on(author_entry(
+                &test_entry_with_value("{\"stuff\":\"test entry value number 1\"}"),
+                None,
+                &context1,
+                &vec![],
+            ))
+            .unwrap()
+            .address();
+        thread::sleep(time::Duration::from_millis(500));
+
+        // Jill publishes another entry
+        context1
+            .block_on(author_entry(
+                &test_entry_with_value("{\"stuff\":\"test entry value number 2\"}"),
+                None,
+                &context1,
+                &vec![],
+            ))
+            .unwrap()
+            .address();
+        thread::sleep(time::Duration::from_millis(500));
+
+        // collect Jills local chain
+        let state = &context1.state().unwrap();
+        let jill_headers: Vec<ChainHeader> = state
+            .agent()
+            .iter_chain()
+            .collect();
+        let header = jill_headers.first().expect("Must be at least one header in chain");
```

Comment:
I’m not sure about ‘first()’ here. Wouldn’t that five the top chain header as the chain iterator walks back the chain starting with the last one? That would mean that this test is not showing that you can get the whole chain, but just the first one..

Reply:
Only think keeping me from approving this PR, @willemolding.

Reply:
I thought that as well and originally had `.last()` but actually that is incorrect. For some reason they are the other way around to what you would expect.

As it is it definitely retrieves the whole chain (from the top two App entry headers plus the AgentId and DNA entries. For clarity I will add an assertion that the retrieved chain has length=5.